### ISSUE TITLE
Provide direct (not nested) path to stimuli

### ISSUE THREAD

**Issue by jankaWIS** [wmvanvliet/pytorch_hmax#1]
Hi,

great repo and effort. I really admire your courage to write HMAX in python. I have a question about loading data in, namely about this part of the code:
https://github.com/wmvanvliet/pytorch_hmax/blob/master/example.py#L18

I know that by default, the `ImageFolder` expects to have nested folders (as stated in [docs](https://pytorch.org/vision/main/generated/torchvision.datasets.ImageFolder.html) or mentioned in this [issue](https://github.com/pytorch/vision/issues/669)) but it's quite clumsy in this case. Eg even if you look at your example, having subfolders for each photo just doesn't look good. Would you have a way how to go around this? Any suggestion on how to provide only a path to all images and not this nested path? I was reading some [discussions](https://discuss.pytorch.org/t/using-only-some-of-subfolders-in-dataset/72323/2) but haven't figured out how to implement it.

---

One more question (I didn't want to open an extra issue for that), shouldn't in https://github.com/wmvanvliet/pytorch_hmax/blob/master/example.py#L28 be 
`batch_size=len(images))` instead of `batch_size=10` (written symbolically)?

Thanks.

**Comment by roshanshrestha01** [comment]
Hi,

I haven't revisit this project for a long time so my take might not be that
accurate.

We need to label images. Here, from ImageFolder we are labeling images from
the parent directory. If you want to have all images in singe folder you
need to label them. If images are label in some way, we can use that to
move them in proper directory structure with simple script.

I would have add a script to properly separate them in proper structure
required by ImageFolder.

In case for adding batch_size, if we provide `len(images)` we are
processing entire images at once. For example, If we have 100 images we are
processing 100 images, provided batch size we are processing 10 images in
10 batch.
Please confirm in documentation about batch_size.

Thank you,
Roshan Shrestha

**Comment by wmvanvliet** [comment]
@roshanshrestha01 is right, (and thank you for responding!), but it is only part of the story.

The original point of this repo was to provide an implementation of HMAX formulated in the PyTorch framework, rather than a desire to have a Python implementation. Since `ImageFolder` is the most obvious manner to import images into this framework, that's what we show in the example. As @roshanshrestha01 pointed out, `ImageFolder` wants images organized in folders as a means to provide class labels for them.

However, the HMAX model in this repo has been pre-trained (the `universal_patch_set.m` file contains the pre-learned weights). As such, it could be used by someone just wanting to have a convenient and fast HMAX implementation to obtain responses for a set of images. For this purpose, no class labels are necessary and @jankaWIS is right to point out that the folder structure then becomes a burden.

In my opinion, anyone with a good understanding of PyTorch should be able to load images into the model in any manner they like. The `example.py` script really is meant just as an example, not some sort of official interface (the official interfaces would be the `torch.nn.Module` interface and the `get_all_layers` method of the `HMAX` class). So, let's see if we can help @jankaWIS out in their desire to load a folder of images without a folder structure by providing a guide to how PyTorch loads and handles input data.

As [the docstring](https://github.com/wmvanvliet/pytorch_hmax/blob/master/hmax.py#L378-L397) states, `HMAX.run_all_layers` expects as input a `torch.tensor` of shape `(batch_size, 1, height, width)`, where `batch_size` is the number of images to process at once, `1` means the images have only one color channel, i.e. they are grayscale, `height` is the height of the images in pixels and `width` is the width of the images in pixels. To address @jankaWIS's second question: HMAX does not care about `batch_size` as such: you will obtain the same result whether you feed in images one at a time or a thousand at a time. Feeding more images through in one go is faster at the expanse of GPU memory. `10` is a decent value in my opinion, but you are free to set it to whatever you like (and your GPU can handle).

So, our goal now it to transform a folder with images (in various file formats, sizes, colors, whatever) into a `torch.tensor` of the correct shape to feed to the model. In the PyTorch framework, image loading is handled by the Pillow library (`PIL`). [`PIL.Image.open`](https://pillow.readthedocs.io/en/stable/reference/Image.html#PIL.Image.open) reads an image. So, loading the images is just a matter of looping through the files in the folder and calling [`PIL.Image.open`](https://pillow.readthedocs.io/en/stable/reference/Image.html#PIL.Image.open) on them.

To transform a `PIL.Image` object to a `torch.tensor`, TorchVision has the concept of "transformers". Take a look at this piece of code in the `example.py` script:

```python
from torchvision import transforms
transform = transforms.Compose([
    transforms.Grayscale(),
    transforms.ToTensor(),
    transforms.Lambda(lambda x: x * 255),
])
```

This defines a function `transform()` that you can call with a `PIL.Image` object as input. It first converts it to grayscale, floating point values from 0 (black) to 1 (white), then to a `torch.tensor` of shape `(1, height, width)`, and finally to values ranging from 0 (black) to 255 (white). This get's us almost all the way there! One thing you may want to add to this transformer pipeline is a [`transforms.Resize`](https://pytorch.org/vision/0.9/transforms.html#torchvision.transforms.Resize) followed by a [`transforms.CenterCrop`](https://pytorch.org/vision/0.9/transforms.html#torchvision.transforms.CenterCrop) to ensure all images end up having the same height and width:

```python
from torchvision import transforms
transform = transforms.Compose([
    transforms.Resize(250),
    transforms.CenterCrop(250),
    transforms.Grayscale(),
    transforms.ToTensor(),
    transforms.Lambda(lambda x: x * 255),
])
```

So that gets us to a bunch of `torch.tensor` objects of size `(1, height, width)`. All that is left to do is stack a batch (say 10) of them using `torch.stack()` to obtain a final tensor of shape `(batch_size, 1, height, width)`. And we're done! You can feed this tensor through the model (either by calling `model()` on it directly, or by using the `model.get_all_layers()` method).

If you want to wrap it all up with a pretty bow, you can implement all of the above by defining a [custom `DataSet` class](https://pytorch.org/tutorials/beginner/basics/data_tutorial.html#creating-a-custom-dataset-for-your-files) so you can keep on using the `DataLoader` in the `example.py` script. That `DataLoader` performs the job of asking the `DataSet` class to load a couple of individual images and then calling `torch.stack` on them. However, it will do this in a separate thread, so that the GPU can start working on one batch of images while the `DataLoader` is loading the next batch of images in the background.

**Comment by jankaWIS** [comment]
First of all, thank you both for your answers. Yes, I am new to pytorch, as you probably both guessed :). And great thanks to @wmvanvliet for your amazing and detailed explanation. You are answering exactly what I had in mind and what I asked about :). To wrap it up in code, this is how you meant it? Feel free to comment and criticise.

### Version 1
```python
import os
import torch
from torchvision import transforms
from PIL import Image

import hmax

# Initialize the model with the universal patch set
print('Constructing model')
model = hmax.HMAX(os.path.join(path_hmax, 'universal_patch_set.mat'))

# Determine whether there is a compatible GPU available
device = torch.device('cuda:0' if torch.cuda.is_available() else 'cpu')

# Define the transform
transform = transforms.Compose([
#    transforms.Resize(250),
#    transforms.CenterCrop(250),
    transforms.Grayscale(),
    transforms.ToTensor(),
    transforms.Lambda(lambda x: x * 255),
])

# maybe this can be done better -> I also didn't add the batch size
tensor_list = []
matlab_img_path = os.path.expanduser('~/Downloads/hmaxMatlab/exampleImages/')

for im in os.listdir(matlab_img_path):
    # load image and transform
    img = Image.open(os.path.join(matlab_img_path, im))
    # append the image
    tensor_list.append(transform(img))

# Create a tensor stack of the images
X = torch.stack(tensor_list)

# Run the model on the example images
print('Running model on', device)
model = model.to(device)
s1, c1, s2, c2 = model.get_all_layers(X.to(device))

print('done')
```

It's missing the batch parameter - that I have not yet figured out. I mean, I guess I could just add another for loop that would go over all the images in steps of `batch_size`, but it looked cumbersome, and I guess that it doesn't help much with any of the loading (unlike the `DataSet`).

### Version 2 (dataset)

```python
import os
import torch
from torchvision import transforms
from torch.utils.data import DataLoader
from PIL import Image

import hmax

class CustomImageDataset():
    def __init__(self, img_dir, img_labels=None, transform=None):
        self.img_labels = os.listdir(img_dir)
        self.img_dir = img_dir
        self.transform = transform

    def __len__(self):
        return len(self.img_labels)

    def __getitem__(self, idx):
        img_path = os.path.join(self.img_dir, self.img_labels[idx])
        image = Image.open(img_path)

        label = self.img_labels[idx]
        if self.transform:
            image = self.transform(image)
        return image, label


    
# Initialize the model with the universal patch set
print('Constructing model')
model = hmax.HMAX(os.path.join(path_hmax, 'universal_patch_set.mat'))

# Determine whether there is a compatible GPU available
device = torch.device('cuda:0' if torch.cuda.is_available() else 'cpu')

# Define transformation to be done on images
transform = transforms.Compose([
#    transforms.Resize(250),
#    transforms.CenterCrop(250),
    transforms.Grayscale(),
    transforms.ToTensor(),
    transforms.Lambda(lambda x: x * 255),
])


# Load folder with example images using the customised dataset class
example_images = CustomImageDataset(
    os.path.expanduser('~/Downloads/hmaxMatlab/exampleImages/'),
    transform=transform,
)

# A dataloader that will run through all example images in one batch
dataloader = DataLoader(example_images, batch_size=10)

# Run the model on the example images
print('Running model on', device)
model = model.to(device)
for X, y in dataloader:
    s1, c1, s2, c2 = model.get_all_layers(X.to(device))
    
print('done')    
```


Btw if you wanted to use any of this in your example(s), feel free. I don't know your intentions with this repository.

I think I have everything now, and I could close the issue, but I would leave it open in case you'd have more suggestions or comments. But feel free to close it :). And thanks again for your help.

**Comment by wmvanvliet** [comment]
Looks great! You totally got it.

**Comment by roshanshrestha01** [comment]
Great explanation by Marijn van Vliet and properly implemented by Jan
Kadlec. Thank you for keeping me in a loop.
